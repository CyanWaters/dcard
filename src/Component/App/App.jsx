import React, { useState, useEffect, useCallback } from "react";
import { Input, Layout, Table } from "antd";
import {
  createFromIconfontCN,
  LoadingOutlined,
  GithubOutlined,
  FileOutlined,
} from "@ant-design/icons";

import { InfiniteScroll } from "../InfiniteScroll";
import { searchRepoFromGitHub, debounce } from "../../lib";

import "./styles/index.css";

const DEFAULT_COUNT = 30;
const DEFAULT_PAGE = 1;
const DELAY_INPUT_TIME = 600;
const GITHUB_URL = "https://github.com/CyanWaters";

const IconFont = createFromIconfontCN({
  scriptUrl: "//at.alicdn.com/t/font_8d5l8fzk5b87iudi.js",
});

export const App = () => {
  const [repoName, setRepoName] = useState("");
  const [count, setCount] = useState(DEFAULT_COUNT);
  const [page, setPage] = useState(DEFAULT_PAGE);
  const [loading, setLoading] = useState(false);
  const [dataSource, setDataSource] = useState([]);
  const [errorMessage, setErrorMessage] = useState("");

  const { Header, Footer, Sider, Content } = Layout;

  const disalbeButton = !(repoName && page && count);
  // Table Columns
  const columns = [
    {
      title: "Name",
      dataIndex: "name",
      key: "name",
    },
    {
      title: "URL",
      dataIndex: "url",
      key: "url",
    },
    {
      title: "Latest Updated Time",
      dataIndex: "updated_at",
      key: "updated_at",
    },
  ];

  const onChangeRepoName = useCallback((e) => {
    debounce(() => setRepoName(e?.target?.value), DELAY_INPUT_TIME)();
  }, []);
  const onChangeCount = useCallback((e) => {
    debounce(() => setCount(e?.target?.value), DELAY_INPUT_TIME)();
  }, []);
  const onChangePage = useCallback((e) => {
    debounce(() => setPage(e?.target?.value), DELAY_INPUT_TIME)();
  }, []);

  const getRepo = async (cleanDataSource = false) => {
    setLoading(true);
    const {
      items,
      totalCount,
      error = false,
      errorMessage: err = "",
    } = await searchRepoFromGitHub(repoName, count, page);
    if (error) {
      setErrorMessage(err);
    } else {
      const data = items.map((item, index) => {
        const obj = {
          key: index,
          name: item?.name,
          url: item?.url,
          updated_at: item?.updated_at,
        };
        return obj;
      });
      setDataSource(cleanDataSource ? data : dataSource.concat(data));
      setErrorMessage("");
    }
    setLoading(false);
  };
  const loadMore = useCallback(() => {
    setPage(page + 1);
  }, [page]);
  const tableComponenet = (
    <Table
      className="repo-table"
      dataSource={dataSource}
      columns={columns}
      pagination={false}
    />
  );

  useEffect(() => {
    if (!disalbeButton) {
      // detect repoName is the same or not
      getRepo(true);
    }
  }, [repoName]);
  useEffect(() => {
    if (!disalbeButton) {
      getRepo();
    }
  }, [page, count]);

  return (
    <>
      <div className="app">
        <Layout>
          <Header className="header">
            <a className="link" href={GITHUB_URL} target="popup">
              <GithubOutlined /> GitHub
            </a>
            <div className="float-right link">
              <IconFont type="icon-tuichu" />
            </div>
            <div className="clear-float" />
          </Header>
          <Content>
            <div>
              <p className="field-name">Repoitory Name:</p>
              <Input
                data-testid="repoName"
                placeholder="Repo Name"
                onChange={onChangeRepoName}
              />
            </div>
            <div>
              <p className="field-name"> Per Page Count:</p>
              <Input
                type="number"
                placeholder="Count"
                min={1}
                onChange={onChangeCount}
                defaultValue={count}
              />
            </div>
            <div>
              <p className="field-name"> Page:</p>
              <Input
                type="number"
                placeholder="Page"
                min={1}
                onChange={onChangePage}
                value={page}
                disabled
              />
            </div>
            {/* Search Button */}
            {/* <div className="button-align-center">
            <Button type="primary" disabled={disalbeButton} onClick={getRepo}>
              Search
            </Button>
          </div> */}
            {errorMessage ? (
              <div className="field-name no-data">
                {`Something went wrong: ${errorMessage}`}
              </div>
            ) : disalbeButton || !dataSource.length ? (
              loading ? (
                <div className="data-loading">
                  <LoadingOutlined />
                </div>
              ) : (
                <div className="field-name no-data">
                  <FileOutlined />
                  No Data
                </div>
              )
            ) : (
              <InfiniteScroll
                key={`${count}+${page}+${loading}`}
                loadMore={loadMore}
                component={tableComponenet}
                loading={loading}
              />
            )}
          </Content>
          {/* <Footer>Footer</Footer> */}
        </Layout>
      </div>
    </>
  );
};

export default App;
