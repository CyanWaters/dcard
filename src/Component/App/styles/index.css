.header {
  color: #fff;
  background-color: #24292f;
  line-height: 5em;
  padding-left: 1em;
}

.link {
  color: #fff;
  font-size: 1.5rem;
  margin-right: 0.5em;
}
.float-right {
  float: right;
}
.clear-float {
  clear: both;
}

.content {
  background: #fff;
  line-height: 50em;
}

.field-name {
  display: inline-block;
  margin-right: 1em;
  margin-left: 1em;
}

.repo-table {
  margin-left: 1em;
}

.button-align-center {
  align-items: center;
  display: flex;
  justify-content: center;
}

.data-loading {
  display: block;
  text-align: center;
  background-color: #fff;
  border: 0;
  margin-top: 1em;
}
.no-data {
  font-size: 2.5rem;
}
