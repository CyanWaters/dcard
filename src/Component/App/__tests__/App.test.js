import React from "react";
import TestRenderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { render, fireEvent, act } from "@testing-library/react";
import moxios from "moxios";

import { mockAPIData } from "../__mocks__";
import { App } from "../App";

beforeEach(() => {
  // install the axios mock before each test/describe
  moxios.install();
});

afterEach(() => {
  // uninstall the axios mock after each test/describe
  moxios.uninstall();
});

test("should render well", async () => {
  let renderer;
  await TestRenderer.act(async () => {
    renderer = TestRenderer.create(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );
  });
  expect(renderer.toJSON()).toMatchSnapshot();
});

describe("search functions should be well", () => {
  test("Repo Name was changed", async () => {
    mockAPIData();
    let result;
    await act(async () => {
      result = render(
        <MemoryRouter>
          <App />
        </MemoryRouter>
      );
    });
    const repoName = "react";

    await act(async () => {
      fireEvent.change(result.getByTestId("repoName"), {
        target: {
          value: repoName,
        },
      });
    });

    const searchInput = result.getByTestId("repoName");
    expect(searchInput.value).toEqual(repoName);
    // const url = `https://api.github.com/search/repositories?q=${repoName}&per_page=30&page=1`;
    // mockAPIData(url);

    // to ensure request call again
    // await act(async () => {
    //   result.rerender();
    // });
    const request = moxios.requests.mostRecent();
    expect(request.url).toEqual(url);
  });
});
