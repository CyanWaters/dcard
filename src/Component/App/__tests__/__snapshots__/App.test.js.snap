// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render well 1`] = `
<div
  className="app"
>
  <section
    className="ant-layout"
  >
    <header
      className="ant-layout-header header"
    >
      <a
        className="link"
        href="https://github.com/CyanWaters"
        target="popup"
      >
        <span
          aria-label="github"
          className="anticon anticon-github"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="github"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M511.6 76.3C264.3 76.2 64 276.4 64 523.5 64 718.9 189.3 885 363.8 946c23.5 5.9 19.9-10.8 19.9-22.2v-77.5c-135.7 15.9-141.2-73.9-150.3-88.9C215 726 171.5 718 184.5 703c30.9-15.9 62.4 4 98.9 57.9 26.4 39.1 77.9 32.5 104 26 5.7-23.5 17.9-44.5 34.7-60.8-140.6-25.2-199.2-111-199.2-213 0-49.5 16.3-95 48.3-131.7-20.4-60.5 1.9-112.3 4.9-120 58.1-5.2 118.5 41.6 123.2 45.3 33-8.9 70.7-13.6 112.9-13.6 42.4 0 80.2 4.9 113.5 13.9 11.3-8.6 67.3-48.8 121.3-43.9 2.9 7.7 24.7 58.3 5.5 118 32.4 36.8 48.9 82.7 48.9 132.3 0 102.2-59 188.1-200 212.9a127.5 127.5 0 0138.1 91v112.5c.8 9 0 17.9 15 17.9 177.1-59.7 304.6-227 304.6-424.1 0-247.2-200.4-447.3-447.5-447.3z"
            />
          </svg>
        </span>
         GitHub
      </a>
      <div
        className="float-right link"
      >
        <span
          className="anticon"
          role="img"
        >
          <svg
            aria-hidden="true"
            className=""
            fill="currentColor"
            focusable="false"
            height="1em"
            width="1em"
          >
            <use
              xlinkHref="#icon-tuichu"
            />
          </svg>
        </span>
      </div>
      <div
        className="clear-float"
      />
    </header>
    <main
      className="ant-layout-content"
    >
      <div>
        <p
          className="field-name"
        >
          Repoitory Name:
        </p>
        <input
          className="ant-input"
          data-testid="repoName"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Repo Name"
          type="text"
          value=""
        />
      </div>
      <div>
        <p
          className="field-name"
        >
           Per Page Count:
        </p>
        <input
          className="ant-input"
          min={1}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Count"
          type="number"
          value={30}
        />
      </div>
      <div>
        <p
          className="field-name"
        >
           Page:
        </p>
        <input
          className="ant-input ant-input-disabled"
          disabled={true}
          min={1}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Page"
          type="number"
          value={1}
        />
      </div>
      <div
        className="field-name no-data"
      >
        <span
          aria-label="file"
          className="anticon anticon-file"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="file"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M854.6 288.6L639.4 73.4c-6-6-14.1-9.4-22.6-9.4H192c-17.7 0-32 14.3-32 32v832c0 17.7 14.3 32 32 32h640c17.7 0 32-14.3 32-32V311.3c0-8.5-3.4-16.7-9.4-22.7zM790.2 326H602V137.8L790.2 326zm1.8 562H232V136h302v216a42 42 0 0042 42h216v494z"
            />
          </svg>
        </span>
        No Data
      </div>
    </main>
  </section>
</div>
`;
